reducer/src/main/java/com/graphicsfuzz/reducer/reductionopportunities/LiteralToUniformReductionOpportunity.java

```diff
@@ -16,27 +16,108 @@
 
 public class LiteralToUniformReductionOpportunity
     extends AbstractReductionOpportunity {
 
-  private final Expr literalExpr;
-  private final PipelineInfo pipelineInfo;
+  private final ConstantExpr literalExpr;
+  private final TranslationUnit translationUnit;
+  private final ShaderJob shaderJob;
 
-  LiteralToUniformReductionOpportunity(Expr literalExpr,
-                                       PipelineInfo pipelineInfo,
+  LiteralToUniformReductionOpportunity(ConstantExpr literalExpr,
+                                       TranslationUnit translationUnit,
+                                       ShaderJob shaderJob,
                                        VisitationDepth depth) {
+
     super(depth);
+
     this.literalExpr = literalExpr;
-    this.pipelineInfo = pipelineInfo;
+    this.translationUnit = translationUnit;
+    this.shaderJob = shaderJob;
   }
 
   @Override
   void applyReductionImpl() {
-    // TODO: ...
+
+    final Number numericValue;
+    final String arrayName;
+    final BasicType basicType;
+
+    // If the uniform array doesn't exist in the pipeline info, adds an empty array.
+    if (literalExpr instanceof IntConstantExpr) {
+      arrayName = Constants.INT_LITERAL_UNIFORM_VALUES;
+      basicType = BasicType.INT;
+      numericValue = ((IntConstantExpr) literalExpr).getNumericValue();
+    } else {
+      arrayName = Constants.FLOAT_LITERAL_UNIFORM_VALUES;
+      basicType = BasicType.FLOAT;
+      numericValue = Float.valueOf(((FloatConstantExpr) literalExpr).getValue());
+    }
+
+    if (!shaderJob.getPipelineInfo().hasUniform(arrayName)) {
+      shaderJob.getPipelineInfo().addUniform(arrayName, basicType,
+          Optional.of(0), new ArrayList<>());
+      shaderJob.getPipelineInfo().addUniformBinding(arrayName, false,
+          shaderJob.getPipelineInfo().getUnusedBindingNumber());
+    }
+
+    final int index;
+    final List<Number> values = this.shaderJob.getPipelineInfo().getArgs(arrayName);
+
+    if (values.contains(numericValue)) {
+      index = values.indexOf(numericValue);
+    } else {
+      index = shaderJob.getPipelineInfo().appendValueToUniform(arrayName, numericValue);
+    }
+
+    // Adds the new array declaration to the current translation unit and updates the size
+    // in every translation unit where the declaration is present.
+    final int arraySize = shaderJob.getPipelineInfo().getArgs(arrayName).size();
+    final ArrayInfo arrayInfo = new ArrayInfo(new IntConstantExpr(String.valueOf(arraySize)));
+    final VariableDeclInfo variableDeclInfo = new VariableDeclInfo(arrayName,
+        arrayInfo, null);
+    final VariablesDeclaration arrayDecl = new VariablesDeclaration(
+        new QualifiedType(basicType, Arrays.asList(TypeQualifier.UNIFORM)), variableDeclInfo
+    );
+
+    if (!this.translationUnit.hasUniformDeclaration(arrayName)) {
+      this.translationUnit.addDeclaration(arrayDecl);
+    }
+
+    for (TranslationUnit tu: shaderJob.getShaders()) {
```

Comment:
I would comment before this loop: "We go through each translation unit in the shader job and update its existing array uniform to refer to this new array uniform."

Reply:
Done.